Null Exception on GitHub PURL

### Current Behavior

After updating to v4.10.0 I added PURLs  to the components which maintain their releases on GitHub.
E.g. cjson (`pkg:github/DaveGamble/cJSON` )

Watching the logs of the docker deployed apiserver, I can see that upon saving the component details, an internal analysis task is started, which hits a NullPointerException pretty quickly

The one strange thing that I noticed is: the purl is converted into lowercase! See the log below, and also if one opens the component details, the PURL is displayed lowercase. Maybe this is the culprit of an empty response and that causes a NullException because of a missing check? (Most probably not, otherwise it would succeed with the PURL which is all lowercase: `pkg:github/cesanta/mongoose`)

Note: I tried with both on and off for "Authentication Required" for the GitHub Repository Admin settings

```
dependency-track_apiserver  | 2023-12-13 14:29:11,387 INFO [InternalAnalysisTask] Starting internal analysis task
dependency-track_apiserver  | 2023-12-13 14:29:11,387 INFO [InternalAnalysisTask] Analyzing 1 component(s)
dependency-track_apiserver  | 2023-12-13 14:29:11,404 INFO [InternalAnalysisTask] Internal analysis complete
dependency-track_apiserver  | 2023-12-13 14:29:11,408 INFO [RepositoryMetaAnalyzerTask] Performing component repository metadata analysis against 1 components
dependency-track_apiserver  | 2023-12-13 14:29:11,412 WARN [CacheStampedeBlocker] An error occurred while populating cache repositoryMetaCache for key pkg:github/davegamble/cjson : java.lang.NullPointerException
dependency-track_apiserver  | org.dependencytrack.exception.MetaAnalyzerException: java.lang.NullPointerException
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.GithubMetaAnalyzer.analyze(GithubMetaAnalyzer.java:142)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.RepositoryMetaAnalyzerTask.analyze(RepositoryMetaAnalyzerTask.java:177)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.RepositoryMetaAnalyzerTask.lambda$analyze$0(RepositoryMetaAnalyzerTask.java:121)
dependency-track_apiserver  | 	at io.github.resilience4j.retry.Retry.lambda$decorateCallable$5(Retry.java:237)
dependency-track_apiserver  | 	at io.github.resilience4j.retry.Retry.executeCallable(Retry.java:373)
dependency-track_apiserver  | 	at org.dependencytrack.util.CacheStampedeBlocker.readThroughOrPopulateCache(CacheStampedeBlocker.java:201)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.RepositoryMetaAnalyzerTask.analyze(RepositoryMetaAnalyzerTask.java:126)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.RepositoryMetaAnalyzerTask.inform(RepositoryMetaAnalyzerTask.java:91)
dependency-track_apiserver  | 	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:110)
dependency-track_apiserver  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
dependency-track_apiserver  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
dependency-track_apiserver  | 	at java.base/java.lang.Thread.run(Unknown Source)
dependency-track_apiserver  | Caused by: java.lang.NullPointerException: null
dependency-track_apiserver  | 2023-12-13 14:29:11,413 INFO [RepositoryMetaAnalyzerTask] Completed component repository metadata analysis against 1 components
```



### Steps to Reproduce

1. Find package with releases on github (e.g. https://github.com/DaveGamble/cJSON)
2. Tail the logs of the api server `docker compose logs -f`
3. Create a component with the matching PURL ( `pkg:github/DaveGamble/cJSON`)
4. Save the component (details)

### Expected Behavior

The GitHub PURL is parsed, and the latest version is successfully queried and displayed in the project's component overview table

### Dependency-Track Version

4.10.0

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Database Server Version

15

### Browser

Mozilla Firefox

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

After a restart a maybe helpful hint could be seen in the exception

```
dependency-track_apiserver  | 2023-12-13 14:58:31,430 INFO [InternalAnalysisTask] Starting internal analysis task
dependency-track_apiserver  | 2023-12-13 14:58:31,430 INFO [InternalAnalysisTask] Analyzing 1 component(s)
dependency-track_apiserver  | 2023-12-13 14:58:31,455 INFO [InternalAnalysisTask] Internal analysis complete
dependency-track_apiserver  | 2023-12-13 14:58:31,478 INFO [RepositoryMetaAnalyzerTask] Performing component repository metadata analysis against 1 components
dependency-track_apiserver  | 2023-12-13 14:58:31,553 WARN [CacheStampedeBlocker] An error occurred while populating cache repositoryMetaCache for key pkg:github/davegamble/cjson : java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
dependency-track_apiserver  | org.dependencytrack.exception.MetaAnalyzerException: java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.GithubMetaAnalyzer.analyze(GithubMetaAnalyzer.java:142)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.RepositoryMetaAnalyzerTask.analyze(RepositoryMetaAnalyzerTask.java:177)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.RepositoryMetaAnalyzerTask.lambda$analyze$0(RepositoryMetaAnalyzerTask.java:121)
dependency-track_apiserver  | 	at io.github.resilience4j.retry.Retry.lambda$decorateCallable$5(Retry.java:237)
dependency-track_apiserver  | 	at io.github.resilience4j.retry.Retry.executeCallable(Retry.java:373)
dependency-track_apiserver  | 	at org.dependencytrack.util.CacheStampedeBlocker.readThroughOrPopulateCache(CacheStampedeBlocker.java:201)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.RepositoryMetaAnalyzerTask.analyze(RepositoryMetaAnalyzerTask.java:126)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.RepositoryMetaAnalyzerTask.inform(RepositoryMetaAnalyzerTask.java:91)
dependency-track_apiserver  | 	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:110)
dependency-track_apiserver  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
dependency-track_apiserver  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
dependency-track_apiserver  | 	at java.base/java.lang.Thread.run(Unknown Source)
dependency-track_apiserver  | Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
dependency-track_apiserver  | 	at java.base/java.util.regex.Matcher.getTextLength(Unknown Source)
dependency-track_apiserver  | 	at java.base/java.util.regex.Matcher.reset(Unknown Source)
dependency-track_apiserver  | 	at java.base/java.util.regex.Matcher.<init>(Unknown Source)
dependency-track_apiserver  | 	at java.base/java.util.regex.Pattern.matcher(Unknown Source)
dependency-track_apiserver  | 	at org.dependencytrack.tasks.repositories.GithubMetaAnalyzer.analyze(GithubMetaAnalyzer.java:109)
dependency-track_apiserver  | 	... 11 common frames omitted
dependency-track_apiserver  | 2023-12-13 14:58:31,553 INFO [RepositoryMetaAnalyzerTask] Completed component repository metadata analysis against 1 components
```

the new part beeing:
```
dependency-track_apiserver  | 2023-12-13 14:58:31,553 WARN [CacheStampedeBlocker] An error occurred while populating cache repositoryMetaCache for key pkg:github/davegamble/cjson : java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
dependency-track_apiserver  | org.dependencytrack.exception.MetaAnalyzerException: java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
```

--- a/src/main/java/org/dependencytrack/tasks/repositories/GithubMetaAnalyzer.java
+++ b/src/main/java/org/dependencytrack/tasks/repositories/GithubMetaAnalyzer.java
@@ -30,7 +30,6 @@ import org.kohsuke.github.GHRepository;
 import org.kohsuke.github.GitHub;
 
 import java.io.IOException;
-import java.util.regex.Pattern;
 
 import static org.apache.commons.lang3.StringUtils.isBlank;
 import static org.apache.commons.lang3.StringUtils.isNotBlank;
@@ -45,9 +44,11 @@ public class GithubMetaAnalyzer extends AbstractMetaAnalyzer {
 
     private static final Logger LOGGER = Logger.getLogger(GithubMetaAnalyzer.class);
 
-    private static final int VERSION_TYPE_RELEASE = 1;
-    private static final int VERSION_TYPE_COMMIT = 2;
-    private static final String VERSION_TYPE_PATTERN = "[a-f0-9]{6,40}";
+    private enum VersionType {
+        RELEASE,
+        COMMIT;
+    }
+    private static final VersionType DEFAULT_VERSION_TYPE = VersionType.RELEASE;
     private static final String REPOSITORY_DEFAULT_URL = "https://github.com";
     private String repositoryUrl;
     private String repositoryUser;
@@ -89,6 +90,27 @@ public class GithubMetaAnalyzer extends AbstractMetaAnalyzer {
     }
 
     /**
+     * Checks whether the pURL version is a release or commit
+     * @param component Component Object
+     * @param repository The GH Repository Object defined by the pURL
+     * @return VersionType
+     * @throws IOException when GitHub API Calls fail
+     */
+    private VersionType get_version_type(final Component component, GHRepository repository) throws IOException {
+        if (component.getPurl().getVersion() == null){
+            LOGGER.debug(String.format("Version is not set, assuming %s", DEFAULT_VERSION_TYPE.name()));
+            return DEFAULT_VERSION_TYPE;
+        }
+        if (repository.getReleaseByTagName(component.getPurl().getVersion()) != null){
+            LOGGER.debug("Version is release");
+            return VersionType.RELEASE;
+        } else {
+            LOGGER.debug("Version is commit");
+            return VersionType.COMMIT;
+        }
+    }
+
+    /**
      * {@inheritDoc}
      */
     public MetaModel analyze(final Component component) {
@@ -104,19 +126,12 @@ public class GithubMetaAnalyzer extends AbstractMetaAnalyzer {
                     github = GitHub.connectAnonymously();
                 }
 
-                Pattern version_pattern = Pattern.compile(VERSION_TYPE_PATTERN);
-                final int version_type;
-                if (version_pattern.matcher(component.getPurl().getVersion()).find()) {
-                    LOGGER.debug("Version is commit");
-                    version_type = VERSION_TYPE_COMMIT;
-                } else {
-                    LOGGER.debug("Version is release");
-                    version_type = VERSION_TYPE_RELEASE;
-                }
-
                 GHRepository repository = github.getRepository(String.format("%s/%s", component.getPurl().getNamespace(), component.getPurl().getName()));
                 LOGGER.debug(String.format("Repos is at %s", repository.getUrl()));
-                if (version_type == VERSION_TYPE_RELEASE) {
+
+                final VersionType version_type = get_version_type(component, repository);
+
+                if (version_type == VersionType.RELEASE) {
                     GHRelease latest_release = repository.getLatestRelease();
                     if (latest_release != null) {
                         meta.setLatestVersion(latest_release.getTagName());
@@ -127,12 +142,14 @@ public class GithubMetaAnalyzer extends AbstractMetaAnalyzer {
                         meta.setPublishedTimestamp(current_release.getPublished_at());
                         LOGGER.debug(String.format("Current version published at: %s", meta.getPublishedTimestamp()));
                     }
-                } else {
+                }
+
+                if (version_type == VersionType.COMMIT) {
                     GHBranch default_branch = repository.getBranch(repository.getDefaultBranch());
                     GHCommit latest_release = repository.getCommit(default_branch.getSHA1());
-                    GHCommit current_release = repository.getCommit(component.getPurl().getVersion());
                     meta.setLatestVersion(latest_release.getSHA1());
                     LOGGER.debug(String.format("Latest version: %s", meta.getLatestVersion()));
+                    GHCommit current_release = repository.getCommit(component.getPurl().getVersion());
                     meta.setPublishedTimestamp(current_release.getCommitDate());
                     LOGGER.debug(String.format("Current version published at: %s", meta.getPublishedTimestamp()));
                 }

--- a/src/test/java/org/dependencytrack/tasks/repositories/GitHubMetaAnalyzerTest.java
+++ b/src/test/java/org/dependencytrack/tasks/repositories/GitHubMetaAnalyzerTest.java
@@ -74,4 +74,46 @@ public class GitHubMetaAnalyzerTest {
         Assert.assertTrue(version_pattern.matcher(metaModel.getLatestVersion()).find());
         Assert.assertNotNull(metaModel.getPublishedTimestamp());
     }
+
+    @Test
+    public void testAnalyzerNoVersion() throws Exception{
+        final var component = new Component();
+        component.setPurl(new PackageURL("pkg:github/CycloneDX/cdxgen"));
+
+        final var analyzer = new GithubMetaAnalyzer();
+        Assert.assertTrue(analyzer.isApplicable(component));
+        Assert.assertEquals(RepositoryType.GITHUB, analyzer.supportedRepositoryType());
+
+        MetaModel metaModel = analyzer.analyze(component);
+        Assert.assertNotNull(metaModel.getLatestVersion());
+        Assert.assertTrue(metaModel.getLatestVersion().startsWith("v"));
+        Assert.assertNull(metaModel.getPublishedTimestamp());
+    }
+    @Test
+    public void testAnalyzerInvalidCommit() throws Exception{
+        final var component = new Component();
+        component.setPurl(new PackageURL("pkg:github/CycloneDX/cdxgen@0000000"));
+
+        final var analyzer = new GithubMetaAnalyzer();
+        Assert.assertTrue(analyzer.isApplicable(component));
+        Assert.assertEquals(RepositoryType.GITHUB, analyzer.supportedRepositoryType());
+
+        MetaModel metaModel = analyzer.analyze(component);
+        Assert.assertNotNull(metaModel.getLatestVersion());
+        Assert.assertNull(metaModel.getPublishedTimestamp());
+    }
+
+        @Test
+    public void testAnalyzerInvalidRelease() throws Exception{
+        final var component = new Component();
+        component.setPurl(new PackageURL("pkg:github/CycloneDX/cdxgen@invalid-release"));
+
+        final var analyzer = new GithubMetaAnalyzer();
+        Assert.assertTrue(analyzer.isApplicable(component));
+        Assert.assertEquals(RepositoryType.GITHUB, analyzer.supportedRepositoryType());
+
+        MetaModel metaModel = analyzer.analyze(component);
+        Assert.assertNotNull(metaModel.getLatestVersion());
+        Assert.assertNull(metaModel.getPublishedTimestamp());
+    }
 }

